### ISSUE TITLE
How do we store metrics per epoch to plot graphs?

### ISSUE THREAD

**Issue by kaundinya5** [google/mediapipe#4880]
### Have I written custom code (as opposed to using a stock example script provided in MediaPipe)

None

### OS Platform and Distribution

Ubuntu

### Python Version

3.10

### Task name (e.g. Image classification, Gesture recognition etc.)

Object detection

### Describe the actual behavior

object_detector.ObjectDetector.create(train_data=train_data, validation_data=validation_data, options=options) does not provide any way to store the metrics per epoch

### Describe the expected behaviour

Currently, I am unable to find any documentation on how to store the loss, recall, precision etc per epoch. I would like to plot graphs to visualize how these metrics change per epoch. Is there any way I can achieve this? Is there no way for me to add custom callbacks while creating the model?

### Standalone code/steps you may have used to try to get what you need

```shell
def train_model(train_dataset_path, validation_dataset_path, test_dataset_path, export_dir):
    train_data = object_detector.Dataset.from_coco_folder(train_dataset_path, cache_dir="/tmp/od_data/train")
    validation_data = object_detector.Dataset.from_coco_folder(validation_dataset_path, cache_dir="/tmp/od_data/validation")
    test_data = object_detector.Dataset.from_coco_folder(test_dataset_path, cache_dir="/tmp/od_data/validation")
    print("train data size: ", train_data.size)
    print("validation data size: ", validation_data.size)
    print("test data size: ", validation_data.size)

    spec = object_detector.SupportedModels.MOBILENET_MULTI_AVG_I384
    max_gpus_available = get_max_gpus_available()
    hparams = object_detector.HParams(export_dir=export_dir, num_gpus=max_gpus_available, steps_per_epoch=800, learning_rate=0.6, epochs=100)
    options = object_detector.ObjectDetectorOptions(supported_model=spec, hparams=hparams)

    model = object_detector.ObjectDetector.create(train_data=train_data, validation_data=validation_data, options=options)
```

**Comment by kuaashish** [comment]
Hi @kaundinya5,

We have been prioritising this backlog feature request and plan to implement it soon. In the meantime, you can use the following code snippet to add custom metric functions after calling `create`.

```
metric_functions = model._metric_functions + [tf.keras.metrics.F1Score(), tf.keras.metrics.Precision(), tf.keras.metrics.Recall()]
model._model.compile(
    optimizer=model._optimizer,
    loss=model._loss_function,
    metrics=metric_functions,
)

metrics = model.evaluate(test_data)
print(metrics)
```

Thank you!

**Comment by kuaashish** [comment]
Hi @joezoug,

Could you please look into this issue? Thank you

**Comment by joezoug** [comment]
Hi @kaundinya5, you can use [tensorboard](https://www.tensorflow.org/tensorboard) to visualize metrics by epoch.

Unfortunately the callback for object detection was not implemented yet - so you may need to hack together something by calling [ObjectDetector.\_\_init\_\_](https://github.com/google/mediapipe/blob/master/mediapipe/model_maker/python/vision/object_detector/object_detector.py#L40) separately instead of using the [ObjectDetector.create](https://github.com/google/mediapipe/blob/master/mediapipe/model_maker/python/vision/object_detector/object_detector.py#L64) method, and then adding the tensorboard callback to model._callbacks after you've initialized the object detector model. After adding the callback, you can call `_create_and_train_model` just like how it is done in `ObjectDetector.create`.

You can see how the tensorboard logging is set in ImageClassifier: https://github.com/google/mediapipe/blob/master/mediapipe/model_maker/python/vision/image_classifier/image_classifier.py#L59. There is a util function called [get_default_callbacks](https://github.com/google/mediapipe/blob/master/mediapipe/model_maker/python/core/utils/model_util.py#L36) which will get you a tensorboard logging and model checkpointing callback.

Alternatively, the metrics should be printed every epoch so you can also try to capture the logs and use regex to capture the metric values by epoch.

**Comment by kaundinya5** [comment]
Hi @joezoug Yes, that's what we were doing to add a custom callback in our code. Just wanted to know if there was a better way to do this. Thanks, I'll keep a lookout for an update on the repo.

**Comment by grewe** [comment]
Is there any update as to when tensorboard call back will be supported in API?

Can you provide example of how to call the init and set the Tensorboard call back completely as discussed above?
